File: app/domain/schemas/benchmark.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class BenchmarkUpload(BaseModel):
    source: str | None = None
    task_outcomes: dict | None = None
    interaction_rates: dict | None = None
    segment_labels: dict | None = None
    attention_data: dict | None = None


class BenchmarkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    study_id: uuid.UUID
    source: str | None
    task_outcomes: dict | None
    interaction_rates: dict | None
    segment_labels: dict | None
    attention_data: dict | None
    version: int
    created_at: datetime

File: web/src/lib/api/benchmark.ts
import { ApiError, apiFetch } from "../apiClient";
import type { Benchmark } from "../types";

// GET .../benchmark 404s until one has ever been uploaded for this study
// (app/services/benchmark_service.py has no row to return) — same "not an
// error, just doesn't exist yet" handling as getAudience.
export async function getBenchmark(studyId: string): Promise<Benchmark | null> {
  try {
    return await apiFetch<Benchmark>(`/studies/${studyId}/benchmark`);
  } catch (err) {
    if (err instanceof ApiError && err.code === "NOT_FOUND") return null;
    throw err;
  }
}

// Upserts — re-uploading bumps `version` server-side rather than erroring
// (app/services/benchmark_service.py:upsert).
export function uploadBenchmark(
  studyId: string,
  input: {
    source?: string | null;
    task_outcomes?: Record<string, unknown> | null;
    interaction_rates?: Record<string, unknown> | null;
    segment_labels?: Record<string, unknown> | null;
    attention_data?: Record<string, unknown> | null;
  }
) {
  return apiFetch<Benchmark>(`/studies/${studyId}/benchmark`, { body: input });
}
